common/src/node.rs

```diff
@@ -95,26 +297,94 @@ impl VoxelData {
             VoxelData::Solid(mat) => mat,
         }
     }
+
+    /// Replaces all voxels in the margin of this chunk with the "Void" material. This function is a coarse
+    /// way to ensure that chunks are fully rendered when they need to be, avoiding a rendering bug caused
+    /// by a voxel's surface failing to render because of a margin being solid.
+    /// Until margins are fully implemented, any solid chunk produced by world generation should have its
+    /// margins cleared if it, or any chunk adjacent to it, is edited, since otherwise, the margins could
+    /// be inaccurate.
+    pub fn clear_margin(&mut self, dimension: u8) {
+        let data = self.data_mut(dimension);
+        let lwm = usize::from(dimension) + 2;
+        for z in 0..lwm {
+            for y in 0..lwm {
+                for x in 0..lwm {
+                    if x == 0 || x == lwm - 1 || y == 0 || y == lwm - 1 || z == 0 || z == lwm - 1 {
+                        // The current coordinates correspond to a margin point. Set it to void.
+                        data[x + y * lwm + z * lwm.pow(2)] = Material::Void;
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn is_solid(&self) -> bool {
+        match *self {
+            VoxelData::Dense(_) => false,
+            VoxelData::Solid(_) => true,
+        }
+    }
+
+    /// Returns a `VoxelData` with void margins based on the given `SerializableVoxelData`, or `None` if
+    /// the `SerializableVoxelData` came from a `VoxelData` with the wrong dimension.
+    pub fn from_serializable(serializable: &SerializableVoxelData, dimension: u8) -> Option<Self> {
+        if serializable.voxels.len() != usize::from(dimension).pow(3) {
+            return None;
+        }
+
+        let mut data = vec![Material::Void; (usize::from(dimension) + 2).pow(3)];
+        let mut input_index = 0;
+        for z in 0..dimension {
+            for y in 0..dimension {
+                for x in 0..dimension {
+                    data[Coords([x, y, z]).to_index(dimension)] = serializable.voxels[input_index];
```

Comment:
Might be worth noting here (and below) that we can't use a linear copy because `data` includes margins. I've had to re-derive that inference a couple times already.